### PR TITLE
Use channel arg to disable authority filter

### PR DESCRIPTION
**Pull request by muxi**
Cronet API defined in the way that allows Cronet to be enabled/disabled at runtime, so the original PR #14758 was not right by making the transport selection at build time.

This PR reverts the previous change and introduces a channel arg to disable the authority filter.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
  +0.0%     +64 [None]                                                  +816  +0.0%
   +11%     +80 src/core/ext/filters/http/client_authority_filter.cc     +80   +11%
      +922%     +83 add_client_authority_filter                              +83  +922%

  +0.0%    +144 TOTAL                                                   +896  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/ext/filters/http/client_authority_filter.cc [hunk 1]
```diff
@@ -129,6 +129,16 @@ const grpc_channel_filter grpc_client_authority_filter = {
 
 static bool add_client_authority_filter(grpc_channel_stack_builder* builder,
                                         void* arg) {
+  const grpc_channel_args* channel_args =
+      grpc_channel_stack_builder_get_channel_arguments(builder);
+  const grpc_arg* a = grpc_channel_args_find(
```
Comment (dgquintas) [inline src/core/ext/filters/http/client_authority_filter.cc]: nit: let's use a more descriptive name. How about `disable_client_authority_filter_arg`?

File: src/core/ext/filters/http/client_authority_filter.cc [hunk 2]
```diff
@@ -129,6 +129,16 @@ const grpc_channel_filter grpc_client_authority_filter = {
...
+      channel_args, GRPC_ARG_DISABLE_CLIENT_AUTHORITY_FILTER);
+  if (a != nullptr) {
+    const bool disabled = grpc_channel_arg_get_bool(a, false);
```
Comment (dgquintas) [inline src/core/ext/filters/http/client_authority_filter.cc]: I suggest changing `disabled` to `is_client_authority_filter_disabled`.

File: src/core/ext/filters/http/client_authority_filter.cc [hunk 3]
```diff
@@ -129,6 +129,16 @@ const grpc_channel_filter grpc_client_authority_filter = {
...
+    if (disabled == true) {
```
Comment (dgquintas) [inline src/core/ext/filters/http/client_authority_filter.cc]: you can simply do `return is_client_authority_filter_disabled`

Comment (muxi) [reply src/core/ext/filters/http/client_authority_filter.cc]: No. If `is_client_authority_filter_disabled == false`, we want to continue with `grpc_channel_stack_builder_prepend_filter`.

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
  +0.0%     +64 [None]                                                  +864  +0.0%
   +11%     +80 src/core/ext/filters/http/client_authority_filter.cc     +80   +11%
      +922%     +83 add_client_authority_filter                              +83  +922%

  +0.0%    +144 TOTAL                                                   +944  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] Performance differences noted:
Benchmark                                                               cpu_time    real_time
----------------------------------------------------------------------  ----------  -----------
BM_PumpStreamClientToServer<InProcess>/262144                           +7%         +7%
BM_PumpStreamServerToClient<InProcess>/262144                           +5%         +5%
BM_StreamingPingPong<InProcess, NoOpMutator, NoOpMutator>/262144/2      -4%         -4%
BM_StreamingPingPong<MinInProcess, NoOpMutator, NoOpMutator>/262144/1   -5%         -5%
BM_StreamingPingPongMsgs<MinInProcess, NoOpMutator, NoOpMutator>/32768  -6%         -6%
```

File: src/core/ext/filters/http/client_authority_filter.cc [hunk 4]
```diff
@@ -129,6 +129,17 @@ const grpc_channel_filter grpc_client_authority_filter = {
...
 static bool add_client_authority_filter(grpc_channel_stack_builder* builder,
                                         void* arg) {
+  const grpc_channel_args* channel_args =
+      grpc_channel_stack_builder_get_channel_arguments(builder);
+  const grpc_arg* disable_client_authority_filter_arg = grpc_channel_args_find(
+      channel_args, GRPC_ARG_DISABLE_CLIENT_AUTHORITY_FILTER);
+  if (disable_client_authority_filter_arg != nullptr) {
+    const bool is_client_authority_filter_disabled =
+        grpc_channel_arg_get_bool(disable_client_authority_filter_arg, false);
+    if (is_client_authority_filter_disabled == true) {
```
Comment (dgquintas) [inline src/core/ext/filters/http/client_authority_filter.cc]: nit: `if (is_client_authority_filter_disabled)`

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                            FILE SIZE
 ++++++++++++++ GROWING                                              ++++++++++++++
  +0.0%     +64 [None]                                                  +864  +0.0%
   +11%     +80 src/core/ext/filters/http/client_authority_filter.cc     +80   +11%
      +922%     +83 add_client_authority_filter                              +83  +922%

  +0.0%    +144 TOTAL                                                   +944  +0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by muxi** [comment]
Basic Tests MacOS [opt] - pip install error unrelated.
Basic Tests Multi-language Linux - Ruby flake unrelated
Interop Cloud-to-Cloud Tests - build_docker_java unrelated.

**Pull request merged**